Breeding-Insight/bi-api
# [BI-1606] Allow germplasm records to be in multiple lists

# Description
**Story:** [BI-1606](https://breedinginsight.atlassian.net/jira/software/c/projects/BI/boards/1?modal=detail&selectedIssue=BI-1606)

Instead of using the list name(which may or may not be unique in the future) as the key for the mapping described in the card, a UUID is created for new lists and stored as an entry in the list's external references.

1. Updated Germplasm#constructBrAPIList to add a new external reference with referenceSource set to <BRAPI_BASE_URL>/lists and referenceId set to a randomly generated UUID. 

2. Updated Germplasm#constructBrAPIGermplasm(Program, BreedingMethodEntity, User, boolean, String, Supplier) to accept an argument of UUID germplasmListId

3. Updated Germplasm#constructBrAPIGermplasm(BreedingMethodEntity, User) to accept an argument of UUID germplasmListId

4. Updated the code in Germplasm#constructBrAPIGermplasm(BreedingMethodEntity, User) that currently sets GERMPLASM_IMPORT_ENTRY_NUMBER as follows

    a. Created a Map<String, String> and add an entry where:

        i. key → germplasmListId (from list external reference)
        ii. value → import entry number

    b. Set the map in the additionalInfo of the germplasm with a key of listEntryNumbers

5. Updated BrAPIGermplasmService#processData 

    a. Method now accepts an argument of UUID germplasmListId
    b. updated the logic to set the entry number to either look in the map of import entry numbers for the specific germplasmListId or, for germplasm already created in the old format, look in additionalInfo.importEntryNumber 

6. Updated the sorting of germplasm in BrAPIGermplasmService#exportGermplasmList to fetch the entry number for the list name that’s either within the map of list entry numbers or stored in additionalInfo.importEntryNumber.

7. Updated GermplasmProcessor#process to get the ID for the import list and pass it to Germplasm#constructBrAPIGermplasm.

# Dependencies
biweb develop branch

# Testing
1. Import germplasm with user-supplied entry numbers. 
2. Go to the Germpasm Lists tab and use the dev tools console to inspect the list object in the response and verify there is an external reference entry for /lists and note the UUID stored as the refernceID.
3. Then view all imported germplasm and use the dev tools console to inspect the germplasm objects in the response from bi-api. Verify the additionalInfo field "listEntryNumbers" exists and the value is a Map with the imported list ID from step 2 as the key and entry number as value.

4. Import germplasm without user-supplied entry numbers and repeat process above.

5. Download the imported germplasm and verify the exported file records match what is stored.
6. Download a list of previously imported germplasm that store entry number values in additionalInfo.imortEntryNumber and verify the exported file records match what is stored.

# Checklist:

- [x] I have run TAF: [#136](https://github.com/Breeding-Insight/taf/actions/runs/3526754134)


[BI-1606]: https://breedinginsight.atlassian.net/browse/BI-1606?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
